caffe2/utils/Array.h

```diff
@@ -0,0 +1,316 @@
+/**
+* This file is based on the std::array implementation of libstdc++ at
+* https://gcc.gnu.org/onlinedocs/gcc-7.1.0/libstdc++/api/a01056_source.html
```

Comment:
It would be good to have a comment at this file explaining when to use this Array over `std::array`.

Reply:
In particular, the nightmare scenario is when you have to convert to and back from `c10::array` and `std::array`. Also, I remember we didn't need this when we were running in C++14, so it would also good to have an explanation in code why we need our own reimplementation for C++11.